classes/Hook/HookActionCartSave.php

```diff
@@ -0,0 +1,135 @@
+<?php
+/**
+ * 2007-2020 PrestaShop and Contributors
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Academic Free License 3.0 (AFL-3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/AFL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2020 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/AFL-3.0 Academic Free License 3.0 (AFL-3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\Module\Ps_Googleanalytics\Hooks;
+
+use Configuration;
+use Context;
+use PrestaShop\Module\Ps_Googleanalytics\Handler\GanalyticsDataHandler;
+use PrestaShop\Module\Ps_Googleanalytics\Wrapper\ProductWrapper;
+use Product;
+use Ps_Googleanalytics;
+use Tools;
+use Validate;
+
+class HookActionCartSave implements HookInterface
+{
+    private $module;
+
+    /**
+     * @var Context
+     */
+    private $context;
+
+    public function __construct(Ps_Googleanalytics $module, Context $context)
+    {
+        $this->module = $module;
+        $this->context = $context;
+    }
+
+    /**
+     * run
+     *
+     * @return void
+     */
+    public function run()
+    {
+        if (!isset($this->context->cart)) {
+            return;
+        }
+
+        if (!Tools::getIsset('id_product')) {
+            return;
+        }
+
+        $cart = [
+            'controller' => Tools::getValue('controller'),
+            'addAction' => Tools::getValue('add') ? 'add' : '',
+            'removeAction' => Tools::getValue('delete') ? 'delete' : '',
+            'extraAction' => Tools::getValue('op'),
+            'qty' => (int) Tools::getValue('qty', 1),
+        ];
+
+        $cartProducts = $this->context->cart->getProducts();
+        if (!empty($cartProducts)) {
+            foreach ($cartProducts as $cartProduct) {
+                if ($cartProduct['id_product'] == Tools::getValue('id_product')) {
+                    $addProduct = $cartProduct;
```

Comment:
I think you can `break` the loop to go faster

Reply:
well see :)